{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eda952ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import sys\n",
    "import numpy as np\n",
    "\n",
    "#Redo Figure 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0277811",
   "metadata": {},
   "outputs": [],
   "source": [
    "def RedoFig4C_D(TEtool):\n",
    "    #Redo the figure with the motifs \n",
    "    df = pd.read_csv(\"Data_motifs_new.csv\")\n",
    "\n",
    "    df[[\"ID\", \"Other\"]] =  df[\"unique_id\"].str.split(\"::\", n=1, expand=True)\n",
    "    df.loc[df[\"type\"]==\"transcript\",\"type\"] = \"Transcript\"\n",
    "    df[\"ID\"] = df[\"ID\"]+\"_\" + df[\"population\"] +\"_\" + df[\"type\"]\n",
    "\n",
    "    tr = pd.read_csv(f\"Fig1A_Dataframe_{TEtool}_allseqs.csv\")\n",
    "    x = pd.merge(df, tr, on=\"ID\", how = \"left\")\n",
    "\n",
    "    x.loc[x[\"type\"]!=\"Transcript\",\"TEpresence\"] =x[\"type\"]\n",
    "    x.to_csv(f\"{TEtool}_dataFig4_motif.csv\")\n",
    "    fig, axs = plt.subplots(ncols=2, figsize=(11,4))\n",
    "    ax = sns.pointplot(\n",
    "        data=x, x=\"TEpresence\", y=\"number_motifs_per_transcript\", hue = \"TEpresence\",legend = None, \n",
    "        errorbar=(\"ci\", 100), capsize=.4, color=\".5\", linestyle=\"none\", marker=\"D\",order = [\"dn_no_TE\", \"dn_overlap_TE\", \"dn_TE\", \"gene\", \"Intergenic\"], hue_order = [\"dn_no_TE\", \"dn_overlap_TE\", \"dn_TE\", \"gene\", \"Intergenic\"],ax = axs[0],palette = [\"darkgreen\", \"green\", \"lightgreen\",\"orange\",\"blue\"])\n",
    "    ax.set(title = \"Low identity score\")\n",
    "    ax.set_ylabel(\"# Motif\")\n",
    "    ax.set_xlabel(\" \")\n",
    "\n",
    "    ax = sns.pointplot(\n",
    "        data=x, x=\"TEpresence\", y=\"number_motifs_high_per_transcript\",hue = \"TEpresence\", legend = None, \n",
    "        errorbar=(\"ci\", 100), capsize=.4,\n",
    "        color=\".5\", linestyle=\"none\", marker=\"D\", order = [\"dn_no_TE\", \"dn_overlap_TE\", \"dn_TE\", \"gene\", \"Intergenic\"],hue_order = [\"dn_no_TE\", \"dn_overlap_TE\", \"dn_TE\", \"gene\", \"Intergenic\"],ax = axs[1],palette = [\"darkgreen\", \"green\", \"lightgreen\",\"orange\",\"blue\"])\n",
    "    ax.set(title = \"High identity score\")\n",
    "    ax.set_ylabel(\"# Motif\")\n",
    "    ax.set_xlabel(\" \")\n",
    "    plt.tight_layout()\n",
    "\n",
    "    plt.suptitle(f\"TEtool = {TEtool}\", fontsize=12)\n",
    "    plt.savefig(f\"{TEtool}4AB.jpg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c551b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def RedoFig4A_B(TEtool):\n",
    "    #Redo the core figure\n",
    "    df = pd.read_csv(\"Data_core_new.csv\")\n",
    "    df[[\"ID\", \"Other\"]] =  df[\"unique_id\"].str.split(\"::\", n=1, expand=True)\n",
    "    df.loc[df[\"type\"]==\"transcript\",\"type\"] = \"Transcript\"\n",
    "    df[\"ID\"] = df[\"ID\"]+\"_\" + df[\"population\"] +\"_\" + df[\"type\"]\n",
    "    print(df)\n",
    "\n",
    "    tr = pd.read_csv(f\"Fig1A_Dataframe_{TEtool}_allseqs.csv\")\n",
    "    x = pd.merge(df, tr, on=\"ID\", how = \"left\")\n",
    "\n",
    "    x.loc[x[\"type\"]!=\"Transcript\",\"TEpresence\"] = x[\"type\"]\n",
    "    print(df)\n",
    "    x.to_csv(f\"{TEtool}_dataFig4_core.csv\")\n",
    "    fig, axs = plt.subplots(ncols=2, figsize=(11,4))\n",
    "    ax = sns.pointplot(\n",
    "        data=x, x=\"TEpresence\", y=\"number_core_per_transcript\", hue = \"TEpresence\",legend = None,\n",
    "        errorbar=(\"ci\", 100), capsize=.4,\n",
    "        color=\".5\", linestyle=\"none\", marker=\"D\", hue_order = [\"dn_no_TE\", \"dn_overlap_TE\", \"dn_TE\", \"gene\", \"Intergenic\"],ax = axs[0],palette = [\"darkgreen\", \"green\", \"lightgreen\",\"orange\",\"blue\"])\n",
    "    ax.set(title = \"Low identity score\")\n",
    "    ax.set_ylabel(\"# Core\")\n",
    "    ax.set_xlabel(\" \")\n",
    "\n",
    "    ax = sns.pointplot(\n",
    "        data=x, x=\"TEpresence\", y=\"number_core_high_per_transcript\",hue = \"TEpresence\",legend = None,\n",
    "        errorbar=(\"ci\", 100), capsize=.4, color=\".5\", linestyle=\"none\", marker=\"D\", hue_order = [\"dn_no_TE\", \"dn_overlap_TE\", \"dn_TE\", \"gene\", \"Intergenic\"],ax = axs[1],palette = [\"darkgreen\", \"green\", \"lightgreen\",\"orange\",\"blue\"])\n",
    "    ax.set(title = \"High identity score\")\n",
    "    ax.set_ylabel(\"# Core\")\n",
    "    ax.set_xlabel(\" \")\n",
    "    plt.tight_layout()\n",
    "\n",
    "    plt.suptitle(f\"TEtool = {TEtool}\", fontsize=12)\n",
    "    plt.savefig(f\"{TEtool}4CD.jpg\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
